File: ios/SecureChat/SecureChatNetwork/Sources/SignalMessage.swift
import Foundation

/**
 * Signaling sunucusu üzerinden iletilen mesaj tipleri.
 * WebSocket bağlantısı üzerinden JSON olarak serialize/deserialize edilir.
 *
 * Her mesaj tipi farklı bir amaca hizmet eder:
 * - SdpOffer/SdpAnswer: WebRTC bağlantı kurulumu
 * - IceCandidate: NAT traversal için aday bilgisi
 * - EncryptedMessage: Signal Protocol ile şifrelenmiş mesaj
 * - PreKeyBundleMessage: İlk bağlantı için anahtar değişimi
 * - CallControl: Arama yaşam döngüsü kontrol mesajları
 */

// MARK: - Base Protocol

/// Tüm signaling mesajlarının ortak arayüzü
public protocol SignalMessageProtocol: Codable {
    var senderId: String { get }
    var recipientId: String { get }
    var timestamp: Int64 { get }
    var messageType: String { get }
}

// MARK: - Message Types

/// WebRTC SDP Offer — bağlantı başlatma isteği
public struct SdpOfferMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "sdp_offer"
    public let sdp: String
    public let callType: CallType

    public init(senderId: String, recipientId: String, timestamp: Int64, sdp: String, callType: CallType) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.sdp = sdp
        self.callType = callType
    }
}

/// WebRTC SDP Answer — bağlantı kabul yanıtı
public struct SdpAnswerMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "sdp_answer"
    public let sdp: String

    public init(senderId: String, recipientId: String, timestamp: Int64, sdp: String) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.sdp = sdp
    }
}

/// WebRTC ICE Candidate — NAT traversal için aday bilgisi
public struct IceCandidateMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "ice_candidate"
    public let candidate: String
    public let sdpMid: String?
    public let sdpMLineIndex: Int32

    public init(senderId: String, recipientId: String, timestamp: Int64, candidate: String, sdpMid: String?, sdpMLineIndex: Int32) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.candidate = candidate
        self.sdpMid = sdpMid
        self.sdpMLineIndex = sdpMLineIndex
    }
}

/// Signal Protocol ile şifrelenmiş mesaj zarfı
public struct EncryptedMessageSignal: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "encrypted_message"
    public let envelope: String // Base64 encoded EncryptedEnvelope

    public init(senderId: String, recipientId: String, timestamp: Int64, envelope: String) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.envelope = envelope
    }
}

/// X3DH key agreement için PreKey bundle mesajı
public struct PreKeyBundleMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "prekey_bundle"
    public let bundle: String // Serialized PreKey bundle

    public init(senderId: String, recipientId: String, timestamp: Int64, bundle: String) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.bundle = bundle
    }
}

/// Arama kontrol mesajı (çalma, kabul, red, kapatma, meşgul)
public struct CallControlMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "call_control"
    public let action: CallAction

    public init(senderId: String, recipientId: String, timestamp: Int64, action: CallAction) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.action = action
    }
}

/// WebSocket üzerinden iletilen ses verisi
public struct AudioDataMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "audio_data"
    public let data: String // Base64 encoded PCM ses verisi

    public init(senderId: String, recipientId: String, timestamp: Int64, data: String) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.data = data
    }
}

/// WebSocket üzerinden iletilen video frame verisi
public struct VideoDataMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "video_data"
    public let data: String // Base64 encoded JPEG frame
    public let width: Int32
    public let height: Int32

    public init(senderId: String, recipientId: String, timestamp: Int64, data: String, width: Int32, height: Int32) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.data = data
        self.width = width
        self.height = height
    }
}

/// WebSocket üzerinden dosya transferi
public struct FileTransferMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "file_transfer"
    public let fileName: String
    public let mimeType: String
    public let fileSize: Int64
    public let data: String // Base64 encoded dosya içeriği

    public init(senderId: String, recipientId: String, timestamp: Int64, fileName: String, mimeType: String, fileSize: Int64, data: String) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.fileName = fileName
        self.mimeType = mimeType
        self.fileSize = fileSize
        self.data = data
    }
}

/// Grup yönetimi bildirimleri
public struct GroupNotificationMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "group_notification"
    public let groupId: String
    public let groupName: String
    public let action: GroupAction
    public let groupMembers: [String] // Grup üye listesi (tam liste)
    public let targetMemberId: String? // ADD_MEMBER/REMOVE_MEMBER için hedef üye

    public init(senderId: String, recipientId: String, timestamp: Int64, groupId: String, groupName: String, action: GroupAction, groupMembers: [String], targetMemberId: String? = nil) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.groupId = groupId
        self.groupName = groupName
        self.action = action
        self.groupMembers = groupMembers
        self.targetMemberId = targetMemberId
    }
}

/// Mesaj iletim/okundu bilgisi (delivery receipt)
public struct DeliveryReceiptMessage: SignalMessageProtocol {
    public let senderId: String
    public let recipientId: String
    public let timestamp: Int64
    public let messageType = "delivery_receipt"
    public let messageId: String
    public let status: String // "DELIVERED" veya "READ"

    public init(senderId: String, recipientId: String, timestamp: Int64, messageId: String, status: String) {
        self.senderId = senderId
        self.recipientId = recipientId
        self.timestamp = timestamp
        self.messageId = messageId
        self.status = status
    }
}

// MARK: - Supporting Enums

/// Arama türü — ses veya video
public enum CallType: String, Codable, CaseIterable {
    case voice = "VOICE"
    case video = "VIDEO"
}

/// Arama kontrol aksiyonları
public enum CallAction: String, Codable, CaseIterable {
    case ringing = "RINGING"
    case accept = "ACCEPT"
    case reject = "REJECT"
    case hangup = "HANGUP"
    case busy = "BUSY"
}

/// Grup yönetimi aksiyonları
public enum GroupAction: String, Codable, CaseIterable {
    case createGroup = "CREATE_GROUP"
    case addMember = "ADD_MEMBER"
    case removeMember = "REMOVE_MEMBER"
    case leaveGroup = "LEAVE_GROUP"
    case updateGroupName = "UPDATE_GROUP_NAME"
}

// MARK: - Message Factory

/// SignalMessage oluşturucu sınıfı
public struct SignalMessageFactory {

    /// JSON string'den SignalMessage decode eder
    public static func decodeMessage(from jsonString: String) throws -> SignalMessageProtocol {
        guard let jsonData = jsonString.data(using: .utf8) else {
            throw SignalMessageError.invalidJSON
        }

        // İlk olarak type field'ını okuyan basit struct
        let typeContainer = try JSONDecoder().decode(MessageTypeContainer.self, from: jsonData)

        switch typeContainer.messageType {
        case "sdp_offer":
            return try JSONDecoder().decode(SdpOfferMessage.self, from: jsonData)
        case "sdp_answer":
            return try JSONDecoder().decode(SdpAnswerMessage.self, from: jsonData)
        case "ice_candidate":
            return try JSONDecoder().decode(IceCandidateMessage.self, from: jsonData)
        case "encrypted_message":
            return try JSONDecoder().decode(EncryptedMessageSignal.self, from: jsonData)
        case "prekey_bundle":
            return try JSONDecoder().decode(PreKeyBundleMessage.self, from: jsonData)
        case "call_control":
            return try JSONDecoder().decode(CallControlMessage.self, from: jsonData)
        case "audio_data":
            return try JSONDecoder().decode(AudioDataMessage.self, from: jsonData)
        case "video_data":
            return try JSONDecoder().decode(VideoDataMessage.self, from: jsonData)
        case "file_transfer":
            return try JSONDecoder().decode(FileTransferMessage.self, from: jsonData)
        case "group_notification":
            return try JSONDecoder().decode(GroupNotificationMessage.self, from: jsonData)
        case "delivery_receipt":
            return try JSONDecoder().decode(DeliveryReceiptMessage.self, from: jsonData)
        default:
            throw SignalMessageError.unknownMessageType
        }
    }

    /// SignalMessage'ı JSON string'e encode eder
    public static func encodeMessage(_ message: SignalMessageProtocol) throws -> String {
        let jsonData: Data

        switch message {
        case let msg as SdpOfferMessage:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as SdpAnswerMessage:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as IceCandidateMessage:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as EncryptedMessageSignal:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as PreKeyBundleMessage:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as CallControlMessage:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as AudioDataMessage:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as VideoDataMessage:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as FileTransferMessage:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as GroupNotificationMessage:
            jsonData = try JSONEncoder().encode(msg)
        case let msg as DeliveryReceiptMessage:
            jsonData = try JSONEncoder().encode(msg)
        default:
            throw SignalMessageError.unknownMessageType
        }

        guard let jsonString = String(data: jsonData, encoding: .utf8) else {
            throw SignalMessageError.encodingFailed
        }

        return jsonString
    }
}

// MARK: - Supporting Types

/// MessageType field'ını okumak için kullanılan container
private struct MessageTypeContainer: Codable {
    let messageType: String
}

/// SignalMessage hataları
public enum SignalMessageError: Error {
    case invalidJSON
    case unknownMessageType
    case encodingFailed
}

File: network/src/main/java/com/securechat/network/SignalMessage.kt
package com.securechat.network

import com.securechat.network.model.CallAction
import com.securechat.network.model.CallType
import com.securechat.network.model.GroupAction
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Signaling sunucusu uzerinden iletilen mesaj tipleri.
 * WebSocket baglantisi uzerinden JSON olarak serialize/deserialize edilir.
 *
 * Her mesaj tipi farkli bir amaca hizmet eder:
 * - SdpOffer/SdpAnswer: WebRTC baglanti kurulumu
 * - IceCandidate: NAT traversal icin aday bilgisi
 * - EncryptedMessage: Signal Protocol ile sifrelenmis mesaj
 * - PreKeyBundleMessage: Ilk baglanti icin anahtar degisimi
 * - CallControl: Arama yasam dongusu kontrol mesajlari
 */
@Serializable
sealed class SignalMessage {
    abstract val senderId: String
    abstract val recipientId: String
    abstract val timestamp: Long

    /** WebRTC SDP Offer — baglanti baslatma istegi. */
    @Serializable
    @SerialName("sdp_offer")
    data class SdpOffer(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val sdp: String,
        val callType: CallType
    ) : SignalMessage()

    /** WebRTC SDP Answer — baglanti kabul yaniti. */
    @Serializable
    @SerialName("sdp_answer")
    data class SdpAnswer(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val sdp: String
    ) : SignalMessage()

    /** WebRTC ICE Candidate — NAT traversal icin aday bilgisi. */
    @Serializable
    @SerialName("ice_candidate")
    data class IceCandidate(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val candidate: String,
        val sdpMid: String?,
        val sdpMLineIndex: Int
    ) : SignalMessage()

    /** Signal Protocol ile sifrelenmis mesaj zarfi. */
    @Serializable
    @SerialName("encrypted_message")
    data class EncryptedMessage(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val envelope: String
    ) : SignalMessage()

    /** X3DH key agreement icin PreKey bundle mesaji. */
    @Serializable
    @SerialName("prekey_bundle")
    data class PreKeyBundleMessage(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val bundle: String
    ) : SignalMessage()

    /** Arama kontrol mesaji (calma, kabul, red, kapatma, mesgul). */
    @Serializable
    @SerialName("call_control")
    data class CallControl(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val action: CallAction,
        // Server ACK icin benzersiz ID — client retry icin kullanir.
        val messageId: String? = null,
        // Grup aramasinda: koordinator HANGUP gonderirken doldurulur — server SFU room
        // ve GroupCallSessionStore temizligi icin kullanir.
        val groupId: String? = null
    ) : SignalMessage()

    /**
     * Server'dan client'a — HANGUP/REJECT/BUSY/ACCEPT ulastiginin onayi.
     * Client retry'i durdurmak icin bu mesaji bekler.
     */
    @Serializable
    @SerialName("call_control_ack")
    data class CallControlAck(
        override val senderId: String = "server",
        override val recipientId: String = "",
        override val timestamp: Long,
        val messageId: String,
        val action: String
    ) : SignalMessage()

    /**
     * WebSocket uzerinden iletilen ses verisi.
     * WebRTC yerine dogrudan signaling kanali uzerinden ses akisi icin kullanilir.
     * PCM ses verisi Base64 ile encode edilir.
     */
    @Serializable
    @SerialName("audio_data")
    data class AudioData(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val data: String // Base64 encoded PCM ses verisi
    ) : SignalMessage()

    /**
     * WebSocket uzerinden iletilen video frame verisi.
     * Kamera goruntusu JPEG olarak sikistirilir ve Base64 ile encode edilir.
     * Dusuk cozunurluk (320x240) ve dusuk JPEG kalitesi (%25) ile bant genisligi optimize edilir.
     * Hedef: 5-8 FPS.
     */
    @Serializable
    @SerialName("video_data")
    data class VideoData(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val data: String, // Base64 encoded JPEG frame
        val width: Int,
        val height: Int
    ) : SignalMessage()

    /**
     * WebSocket uzerinden dosya transferi.
     * Kucuk dosyalar tek parca, buyuk dosyalar chunk'lar halinde gonderilir.
     * Resim, video, belge (pdf, word, zip, txt vb.) destekler.
     *
     * Chunk sistemi: transferId ayni olan parcalar birlestirilir.
     * chunkIndex=0'dan baslar, chunkIndex==totalChunks-1 son parcadir.
     * totalChunks=1 ise tek parca (geriye uyumlu).
     */
    @Serializable
    @SerialName("file_transfer")
    data class FileTransfer(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val fileName: String,
        val mimeType: String,
        val fileSize: Long,
        val data: String, // Base64 encoded dosya icerigi (veya chunk)
        val groupId: String? = null,
        val groupName: String? = null,
        val transferId: String? = null, // Chunk'lari eslestirmek icin benzersiz ID
        val chunkIndex: Int = 0, // Bu parcanin sirasi (0-based)
        val totalChunks: Int = 1, // Toplam parca sayisi (1 = tek parca)
        val caption: String? = null, // Medya altyazisi — alici tarafta ayni baloncukta gosterilir
        val isViewOnce: Boolean = false, // Tek gosterimlik medya bayragi
        val originalMessageId: String? = null // Gondericinin orijinal mesaj ID'si (delivery receipt + view-once edit icin)
    ) : SignalMessage()

    /**
     * Grup yonetimi bildirimleri.
     * Grup olusturma, uye ekleme/cıkarma, grup adı degistirme vb. islemler icin.
     */
    @Serializable
    @SerialName("group_notification")
    data class GroupNotification(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val groupId: String,
        val groupName: String,
        val action: GroupAction,
        val groupMembers: List<String>, // Grup uye listesi (tam liste)
        val targetMemberId: String? = null // ADD_MEMBER/REMOVE_MEMBER icin hedef uye
    ) : SignalMessage()

    /**
     * Mesaj iletim/okundu bilgisi (delivery receipt).
     * Alici cihaza ulastiginda DELIVERED, sohbet ekrani acildiginda READ gonderilir.
     */
    @Serializable
    @SerialName("delivery_receipt")
    data class DeliveryReceipt(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val messageId: String,
        val status: String // "DELIVERED" veya "READ"
    ) : SignalMessage()

    /**
     * Herkesten mesaj silme bildirimi.
     * Gonderici kendi mesajini herkesten sildiginde karsi tarafa iletilir.
     */
    @Serializable
    @SerialName("message_delete")
    data class MessageDelete(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val messageId: String
    ) : SignalMessage()

    /**
     * Mesaj duzenleme bildirimi.
     * Gonderici kendi mesajini duzenledikten sonra karsi tarafa iletilir.
     * Sadece 15 dakika icinde gonderilmis mesajlar duzenlenebilir.
     */
    @Serializable
    @SerialName("message_edit")
    data class MessageEdit(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val messageId: String,
        val newContent: String
    ) : SignalMessage()

    /**
     * Yazma gostergesi. Kullanici yazmaya basladiginda/biraktiginda karsi tarafa iletilir.
     */
    @Serializable
    @SerialName("typing_indicator")
    data class TypingIndicator(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val isTyping: Boolean
    ) : SignalMessage()

    /**
     * Sureli mesaj zamanlayici ayari.
     * Bir konusmada sureli mesaj modu degistiginde karsi tarafa bildirilir.
     * duration 0 ise sureli mesaj kapali demektir.
     * conversationId: birebir icin senderId, grup icin groupId
     */
    @Serializable
    @SerialName("disappearing_timer")
    data class DisappearingTimer(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val duration: Long, // milisaniye, 0 = kapali
        val conversationId: String = "" // bos ise senderId kullanilir (geriye uyumluluk)
    ) : SignalMessage()

    /**
     * Mesaj reaksiyonu (emoji).
     * Kullanici bir mesaja emoji ile tepki verdiginde karsi tarafa iletilir.
     * remove=true ise mevcut reaksiyon kaldirilir.
     */
    @Serializable
    @SerialName("message_reaction")
    data class MessageReaction(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val messageId: String,
        val emoji: String,
        val remove: Boolean = false,
        val groupId: String? = null
    ) : SignalMessage()

    /**
     * Sunucu tarafinda grup mesaj dagitimi (fan-out).
     * Sender tek mesaj gonderir, sunucu her uye icin ayri mesaj olusturur.
     * recipientPayloads: {userId -> sifrelenmis envelope} seklinde her uyenin payloadi.
     */
    @Serializable
    @SerialName("group_message_fanout")
    data class GroupMessageFanout(
        override val senderId: String,
        override val recipientId: String = "server", // Sunucu isleyecek
        override val timestamp: Long,
        val groupId: String,
        val recipientPayloads: Map<String, String> // userId -> envelope
    ) : SignalMessage()

    /**
     * SFU room olusturuldu bildirimi.
     * 4+ katilimcili grup aramasinda sunucu Janus VideoRoom olusturur ve bu bilgiyi gonderir.
     * Client bu bilgiyle mesh yerine Janus SFU'ya baglanir.
     */
    @Serializable
    @SerialName("sfu_room_created")
    data class SfuRoomCreated(
        override val senderId: String = "server",
        override val recipientId: String = "broadcast",
        override val timestamp: Long,
        val groupId: String,
        val roomId: Long,
        val janusWsUrl: String,
        val apiSecret: String = ""
    ) : SignalMessage()

    /**
     * Sunucu kapanma bildirimi. Graceful shutdown sirasinda gonderilir.
     * Client bu mesaji alinca reconnect oncesi 5sn bekler.
     */
    @Serializable
    @SerialName("server_shutdown")
    data class ServerShutdown(
        override val senderId: String = "server",
        override val recipientId: String = "broadcast",
        override val timestamp: Long,
        val message: String = ""
    ) : SignalMessage()

    /** Cevrimici durum bildirimi. Sunucu tarafindan subscribe olan istemcilere gonderilir. */
    @Serializable
    @SerialName("presence_update")
    data class PresenceUpdate(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val isOnline: Boolean,
        val lastSeen: Long,
        val hideLastSeen: Boolean = false
    ) : SignalMessage()

    /** Belirli bir kullanicinin presence durumuna abone olma istegi. Sunucu tarafindan islenir. */
    @Serializable
    @SerialName("presence_subscribe")
    data class PresenceSubscribe(
        override val senderId: String,
        override val recipientId: String, // izlenmek istenen kullanici
        override val timestamp: Long
    ) : SignalMessage()

    /** Presence aboneligini iptal etme istegi. Sunucu tarafindan islenir. */
    @Serializable
    @SerialName("presence_unsubscribe")
    data class PresenceUnsubscribe(
        override val senderId: String,
        override val recipientId: String, // abonelik iptal edilecek kullanici
        override val timestamp: Long
    ) : SignalMessage()

    /**
     * Grup arama davetiyesi. Arayan tum grup uyelerine gonderir.
     * Her uye icin ayri mesaj gonderilir (recipientId = uye ID'si).
     * Mesh WebRTC: Her katilimci birbiriyle dogrudan PeerConnection kurar.
     */
    @Serializable
    @SerialName("group_call_invite")
    data class GroupCallInvite(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val groupId: String,
        val callType: CallType,
        val callId: String,
        val participants: List<String> // Tum davet edilen katilimcilar (arayan dahil)
    ) : SignalMessage()

    /**
     * Grup aramasina yeni uye katildi bildirimi.
     * Arayan (koordinator) mevcut katilimcilara gonderir.
     * Mevcut katilimcilar yeni uyeye PeerConnection kurar.
     */
    @Serializable
    @SerialName("group_call_member_joined")
    data class GroupCallMemberJoined(
        override val senderId: String,
        override val recipientId: String,
        override val timestamp: Long,
        val groupCallId: String,
        val joinedMemberId: String
    ) : SignalMessage()

    /**
     * Aktif grup aramasina sonradan katilim istegi.
     * Yeni katilan kullanici tarafindan koordinatore gonderilir.
     * Koordinator validate edip onGroupMemberAccepted akisini tetikler.
     */
    @Serializable
    @SerialName("group_call_join_request")
    data class GroupCallJoinRequest(
        override val senderId: String,
        override val recipientId: String, // Koordinator ID'si
        override val timestamp: Long,
        val groupId: String,
        val callId: String,
        val callType: CallType
    ) : SignalMessage()

    /**
     * Grup icin aktif arama durumu sorgusu.
     * Istemci ChatScreen acildiginda sunucuya gonderir; sunucu cevaplar.
     * recipientId = "server" sabit string'i.
     */
    @Serializable
    @SerialName("group_call_status_query")
    data class GroupCallStatusQuery(
        override val senderId: String,
        override val recipientId: String, // "server"
        override val timestamp: Long,
        val groupId: String
    ) : SignalMessage()

    /**
     * Aktif grup aramasi durum cevabi (sunucudan istemciye).
     * isActive=false ise diger alanlar null/bos.
     * mode: "MESH" veya "SFU". SFU modunda sfuRoomId/janusWsUrl/apiSecret doludur.
     */
    @Serializable
    @SerialName("group_call_status_response")
    data class GroupCallStatusResponse(
        override val senderId: String, // "server"
        override val recipientId: String,
        override val timestamp: Long,
        val groupId: String,
        val isActive: Boolean,
        val callId: String? = null,
        val coordinatorId: String? = null,
        val callType: CallType? = null,
        val participants: List<String> = emptyList(),
        val mode: String? = null, // "MESH" veya "SFU"
        val sfuRoomId: Long? = null,
        val janusWsUrl: String? = null,
        val apiSecret: String? = null
    ) : SignalMessage()
}
